src/nginx/grpc_server_call.cc

```diff
@@ -97,6 +97,16 @@ ngx_int_t ngx_esp_write_output(ngx_http_request_t *r, ngx_chain_t *out,
     return rc;
   }
 
+  // If write data is bufferred by SSL, ngx_http_ouput_filter() may return
+  // NGX_AGAIN.
+  // ngx_handle_write_event() will not work since r->connection->write could be
+  // either active or ready. To work around this prolem, change return code to
+  // NGX_OK
+  // to continue the data flow.
+  if (r->connection->write->active || r->connection->write->ready) {
```

Comment:
Use r->connection->buffered to see whether it is SSL buffered. bitwise and against NGX_LOWLEVEL_BUFFERED

Reply:
I think checking write->active or write->ready is better.  
ngx_handle_write_event at http://lxr.nginx.org/source/src/event/ngx_event.c
is checking them and skip adding the event.  In this case, grpc data flow will stop. By changing to return OK, we will continue grpc data flow.

Reply:
No, the checking there (in ngx_event.c) is to check if you're already in the middle of processing a write event, to avoid indefinite loop. You just need to let the flow go if it is buffered at SSL level (or HTTP2), but not when there is a back pressure. See http://lxr.nginx.org/source/src/http/ngx_http_write_filter_module.c at line 313.